### ISSUE TITLE
Inquiry about layer Equalization implementation

### ISSUE THREAD

**Issue by rematchka** [jakc4103/DFQ#29]
Hi,
I am trying to reproduce the same results but in Keras, I have a couple of questions regarding the implementation

1. I was wondering in `_layer_equalization` function, shouldn't equation be `1 / s = (1 / r2) * sqrt(r1 * r2)` instead  of `1 / s = (1 / r1) * sqrt(r1 * r2)` and ` range_1 = 2* torch.max(torch.abs(weight_first_group[ii]))` instead of `range_1 = torch.max(torch.abs(weight_first_group[ii]))`
2. another thing when you create a relation graph, is it for every 2 layers which are connected, even if one nodes have 2 incoming connection, and of types convolution and depthwise convolution only?
3.  Do the scales sometimes reach nan and diff infinity?

Thank you in advance

**Comment by jakc4103** [comment]
1. The equations are actually mathematical equivalent as a result.
2. Conv, depthwise conv, linear layer types can be equalized. And perhaps transposed conv could also be equalized, too.
    Layers have input branching or output branching will not be included while finding relations, and these kind of layers should not be equalized since the current solution of finding scale value did not properly handle branching cases.
3. Yes, so one might need some EPS while calculating division or cutoff the scale in some range for numerical stability.

**Comment by rematchka** [comment]
1.  by linear layer you mean 2 consecutive fully connected layers?
2. in the Relation graph doesn't the bottom represent the outbounds of each layer, therefore the dictionary is the index of out_bound + index of the layer?
3.  Does the shape of conv layer is [output channel, input channel, height, width]?
4. Does the relation graph for mobilenet v2 in each bottle neck includes conv-->depthwise and depthwise--> conv, as i have a problem in the second relation when calculating scales for group numbers?

**Comment by jakc4103** [comment]
1. Layer equalization can be applied to any 2 consecutive conv, deconv, linear layers. In my unserstanding, since the scale is applied on input channel and output channel, the 2 consecutive layers does not need to be of the same type.
2. The relation contains 1 pair of 2 layers that can be equalized. And only the layers met the requirements (target type, no branching) can form a relation.
3. The shape of Conv weight follows PyTorch default, which is [output_channel, input_channel, kernel_h, kernel_w].
4. Yes, bottle neck of mobilenet v2 are included. You might want to check the dimension order because pytorch and keras did not share the same order.

**Comment by rematchka** [comment]
I adjusted the tensor shape thank you.  But problem lies in the relation between depth-wise and Conv layer since depthwise output channel value is 1.
1- isn't linear shape 2D tensor, so how to equalize it with Conv layer of 4D tenor?
2- Does Depth wise layer shape equals [output_channel, input_channel, kernel_h, kernel_w], and output_channel  value is 1, or does output_channel  values equal the value of input_channel?

**Comment by jakc4103** [comment]
1. The scale is applied on 0th dim or 1st dim, so a 2D tensor and 4D tensor should be ok to multiply the scale.
2. Depthwise conv weight is in shape [out_ch, 1, kernel_h, kernel_w].
